ember-client/shared/src/main/scala/org/http4s/ember/client/internal/EmberWSClient.scala

```diff
@@ -45,12 +45,22 @@ object EmberWSClient {
         socketOption <- getSocket(emberClient, httpWSRequest)
         socket <- socketOption.liftTo[F](new RuntimeException("Not an Ember client")).toResource
 
+        closeFrameDeffered <- F.deferred[WebSocketFrame.Close].toResource
+
         clientReceiveQueue <- Queue.bounded[F, WebSocketFrame](100).toResource
         clientSendQueue <- Queue.bounded[F, WebSocketFrame](100).toResource
 
         _ <- socket.reads
           .through(decodeFrames(true))
-          .foreach(clientReceiveQueue.offer(_))
+          .foreach {
+            case f @ WebSocketFrame.Close(_) =>
+              closeFrameDeffered.complete(f)
+              clientReceiveQueue.offer(f)
```

Comment:
If these are `F` actions, you need to chain them: 
```suggestion
              closeFrameDeffered.complete(f) >>
                clientReceiveQueue.offer(f)
```